### ISSUE TITLE
Special method names --> interfaces

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#19074]
The issue proposes to shield special methods from unintended invocations or redefinitions in user code by placing them into interfaces. This proposal is a way to resolve #19038, at least for non-initializer methods. It is due to @dlongnecke-cray, as interpreted by @vasslitvinov.

The running example is the `writeThis` method. To simplify presentation we abbreviate it to `write` and drop the channel argument.

The core of the proposal is to have a predefined interface:
```chpl
interface StdWriteable(Self) {
  proc Self.write();
}
```

The implementations of the method `write` from this interface are special, whereas all other functions named `write` are not.

The default implementation of this special method is provided implicitly for each type that does not implement it explicitly.

### Using a special method

A generic function can invoke this special method through an interface constraint as usual, ex.:
```chpl
proc writeln(arg: StdWriteable) {
  arg.write();     // invokes the special method 'write' on arg
}
```

How about invoking it on a user type? For example, I want to write:
```chpl
record R { }

var r: R;
r.write();  // I want to call the special 'write'
```

Case **(CALL-EXP)**: user code defines `R.write` explicitly and the definition is outside an `implements`.
Then, it can be invoked normally, like a non-special method:

```chpl
record R { }

// CALL-EXP-1: this 'implements' is explicitly written in user code.
// ==> The user declaration of R.write() defines the special method.
// CALL-EXP-2: user code does not include this 'implements' explicitly.
// ==> The user declaration of R.write() defines a non-special method
// that happens to have the same name.
R implements StdWriteable;

proc R.write() { }       

var r: R;
r.write();  // invokes R.write() defined above, whether it is special or not
```

Case **(CALL-IFC-EXP)**: user code defines `R.write` explicitly inside `R implements StdWriteable`.
Case **(CALL-IFC-IMP)**: user code does not define the special `R.write` explicitly.
This means the program contains the following declaration, either explicitly in the code or implicitly added by the compiler:
```chpl
R implements StdWriteable {
  proc R.write() { ... either user code or the default implementation ... }
}
```

How to invoke such `R.write` is independent of its special-ness. Here are proposals of how to do it.

**(CALL-CAST)** - apply a "reinterprete" cast, which does not produce a new value. Note that it does not work as smoothly for multi-argument interfaces.
```chpl
var r: R;
(r: StdWriteable).write(); // invokes the implementation of StdWriteable.write()
                           // determined by the visible 'R implements StdWriteable'
```

**(CALL-OPEN)** - makes the implementations of the interface function(s) from the visible `R implements StdWriteable` directly visible. This idea is due to John Skaller.

```chpl
var r: R;
with StdWriteable(R) {
  r.write();               // invokes the implementation of StdWriteable.write()
                           // determined by the visible 'R implements StdWriteable'
}
```

We can tune the syntax. Or we can use a `use`:
```chpl
var r: R;
use StdWriteable(R);
r.write();                 // invokes the implementation of StdWriteable.write()
                           // determined by the visible 'R implements StdWriteable'
```

### Defining a special method

The rule is that the default implementation of the interface `StdWriteable` is defined implicitly for any type that does not provide it explicitly.

Here are the various possibilities. Assume that these declarations are always present:
```chpl
record R { }
var r: R;
```

```chpl
// case EXP/EXP-OUTSIDE:
R implements StdWriteable;
proc R.write() { } // defines the special method, which can be invoked normally
r.write();         // ok, invokes the user-defined special method
```
```chpl
// case EXP/EXP-INSIDE:
R implements StdWriteable {
  proc R.write() { } // defines the special method
}
r.write();           // error, unless there is another visible declaration
                     // of 'proc R.write()'
```
```chpl
// case EXP/IMP:
R implements StdWriteable;
// ==> The compiler adds an implicit body of this 'implements', as if the user wrote:
R implements StdWriteable {
  proc R.write() { } // the default implementation of the special method
}
```
```chpl
// case IMP/IMP:
// User code does not contain 'R implements StdWriteable' or 'R.write'.
// ==> The compiler adds an implicit 'implements' with the default implementation, as if the user wrote:
R implements StdWriteable {
  proc R.write() { } // the default implementation of the special method
}
```
```chpl
// case IMP/EXP:
proc R.write() { } // defines a NON-special method; it can be invoked normally
r.write();         // ok, invokes the non-special method
// ==> The compiler adds an implicit 'implements' with the default implementation, as in the case IMP/IMP:
R implements StdWriteable {
  proc R.write() { } // the default implementation of the special method
}
```

**Comment by dlongnecke-cray** [comment]
Thanks for interpreting my rambles and making something so easily digestible 😄.

**Comment by mppf** [comment]
I think this is worth considering but I had thought that part of the reason to use the interface strategy is to not reserve the name of the method. But I think this approach as described above does still reserve the name of the method:
  * If `proc R.write` is declared and there is no `implements` clause, then the compiler will use it in its own inferred `R implements StdWriteable`. Doesn't that mean that `proc R.write` can only be used for this purpose, and so is "reserved"?
  
Suppose for a moment that the compiler did not infer `R implements StdWriteable` when there is a `proc R.write`. Then if it is possible to define `proc R.write` that is not part of `StdWriteable`, we can have two functions both named `R.write`. One is defined by the compiler (in `StdWriteable(R)`) and the other by the user. Is this too confusing? If we can figure out a way for it not to be too confusing, then this approach will allow adding new special methods to the language in the future without impacting the behavior of programs that happened to use the same name as a method.

**Comment by bradcray** [comment]
I think I'm missing something about this proposal, maybe due to reading too quickly.  In a sentence or two: In this proposal, if the compiler automatically fulfills the interface for a user-defined type, what is the benefit relative to what we have today?  What new power / namespace safety does it give the user?

**Comment by dlongnecke-cray** [comment]
One of the issues tackled in the OP are different ways to bring interface methods into scope, see (CALL-OPEN) or (CALL-CAST). I think one idea here is that unless you bring the interface methods into scope, neither set of interface methods (either `StdWriteable` or `UserWriteable` will be considered in the set of visible candidates for the call.

**Comment by vasslitvinov** [comment]
@bradcray - the benefit of this proposal is that the user cannot reference the special method accidentally. To use it or to define it they need to do interface things and reference the predefined interface name, here StdWriteable.

> If `proc R.write` is declared and there is no implements clause, then the compiler will use it in its own inferred `R implements StdWriteable`.

My intention is the opposite: no implements clause => user-visible `R.write` is not magic.

> We can have two functions both named `R.write`. One is defined by the compiler (in StdWriteable(R)) and the other by the user. Is this too confusing?

I imagine that a user will be in one of these states:
* They don't know about special methods. Then, for them there is only one `R.write` - their own, or non at all. Easy.
* They know about special methods and they are not interested in defining/calling them. Then, again, there is only their own `R.write`, if they have defined it.
* They want to define or invoke the special `R.write`. Then, they will not write a non-special `R.write`. Again, they got only one R.write.

What do you think?

**Comment by bradcray** [comment]
Thanks for the additional explanations.

> What do you think?

The main thing I think is "too bad we don't have interfaces yet!"  Though even if we did, I think another consideration is how much expertise / education is required for a user to understand this model vs. #19050 (say).

**Comment by vasslitvinov** [comment]
I agree on both accounts.

It is conceivable that we can get interfaces working well enough to support this pattern by 2.0. Accepting this proposal would be a good motivator. :)

**Comment by dlongnecke-cray** [comment]
One thing that I think is really cool about interfaces is that they make it easy to document requirements explicitly in code.

I would like to think that this would make it even easier to explain special methods, because we could have a section in the spec for interfaces powering language features, and then the primers can link to the interface documentation as well.

So for experienced users, we can say "if you want to override the standard `write()` behavior, just go implement the `StdWriteable` interface, documented here". And for less experienced users, we can link to the interface primer. And that should be enough for them to go do what they need to do, and only if they're really curious would they go and read the spec.

**Comment by dlongnecke-cray** [comment]
Illustration of the (CALL-CAST) idea from another language: Rust lets users disambiguate conflicting trait methods at call time using "Fully Qualified Syntax". E.g., `<obj as StdWriteable>.write();`. Looks an awful lot like a cast.

https://doc.rust-lang.org/rust-by-example/trait/disambiguating.html

**Comment by bradcray** [comment]
> Accepting this proposal would be a good motivator. :)

I think that if you haven't had sufficient motivation to get interfaces working before now, you haven't been paying attention.  Let's not pretend we're going to hold interfaces hostage over this issue.

> One thing that hasn't been brought up in this discussion is that using interfaces gives library authors the same flexibility to say, "please implement this interface on your class so I can use it" without stealing a name from the class.

Andrew, I read this statement as saying "interfaces are a valuable feature in general" (which I don't think anybody on the team would disagree with) rather than anything specific to this  naming challenge.  Is that true, or am I missing something?

> As proposed in #19050 with only a single chpl special namespace, library authors could not use that approach without them choosing a "unique" prefix and then we're back to programming in C.

I don't know if I agree with this statement because, to me, the appeal of #19050 relates to the whole "this symbol is so special that the compiler knows about it, looks for it, and will provide implementations of it if appropriate" angle.  I.e., we don't (currently) have any aspirations to extend this same feature to library writers ("You too can write a library for which you can teach the compiler how to add necessary routines to a user's type if it doesn't exist!").  So it's almost like reserving the method name as a reserved word, yet without reserving the legal identifier version of it as a reserved word (i.e., reserving `#hash` doesn't prevent a user from using `hash` as an identifier elsewhere in their program).

And, I hate to be a broken record, but we don't have interfaces today (and I don't honestly believe that we'll have them in time for 2.0 anymore.  Yet we need to resolve this naming question in time for 2.0).

**Comment by aconsroe-hpe** [comment]
> > One thing that hasn't been brought up in this discussion is that using interfaces gives library authors the same flexibility to say, "please implement this interface on your class so I can use it" without stealing a name from the class.
> 
> Andrew, I read this statement as saying "interfaces are a valuable feature in general" (which I don't think anybody on the team would disagree with) rather than anything specific to this naming challenge. Is that true, or am I missing something?

The part specific to the naming challenge is "please implement this interface (in the general sense, whether that is a language construct called `interface` or a documented special naming convention or otherwise) on your class so I can use it" without stealing a name from the class.

Maybe no one would disagree with it but I still thought it was missing from the conversation.

We are dealing with similar in issues in C++ today with compiler/next where we want a variety of types to implement some "special" methods for us, whether stringify, hash (all comes back to hash!), debug dump, serialize/deserialize (in the future) etc. Some of those types we control but others are in `std` so we have the same dance of "I'd like to use a consistent name (or set of names) but I can't assume that my name isn't already used by the class and I don't/can't modify their class"

> You too can write a library for which you can teach the compiler how to add necessary routines to a user's type if it doesn't exist!

I imagine this would be the case though if the interface provided a default implementation that took the type being implemented for. Perhaps some things will remain compiler only, but with reflection you could get pretty far I would think.

**Comment by dlongnecke-cray** [comment]
> The part specific to the naming challenge is "please implement this interface (in the general sense, whether that is a language construct called interface or a documented special naming convention or otherwise) on your class so I can use it" without stealing a name from the class.

To make sure I'm understanding @aconsroe-hpe: I think what Andrew might be bringing up here is that even though making special methods live in their own namespace solves this issue for special methods, we still haven't provided a solution for when methods on user types collide with each other?

E.g., if I am a library writer, one option is to expose a class `BaseClass` which provides methods that the user can override in order to control their behavior when passed into library routines.

But that requires the user to make their type a subtype of `Library1.BaseClass` (locking them into a class type and adding all the overhead of `Library1.BaseClass`), and also locks the user out of using those method names for their own purposes. It also locks the user out of having their type subclass `Library2.BaseClass` so that they can use routines in `Library2`.

On the record side of the track, I think the only option right now would be to expose your library routines as generic, and that still doesn't solve the issue of "this library routine calls `thing.foo()`, but I'd like to define `thing.foo()` for my own use elsewhere".

As Andrew and Brad said, maybe this is all stuff that has been said before (and is part of the reason we want to add interfaces in the first place).

**Comment by bradcray** [comment]
> we still haven't provided a solution for when methods on user types collide with each other?
> ...
> maybe this is all stuff that has been said before (and is part of the reason we want to add interfaces in the first place).

In my mind, this is one of the several reasons that we're hoping to add interfaces.

**Comment by dlongnecke-cray** [comment]
There has been some great discussion about this idea in https://github.com/chapel-lang/chapel/issues/19050, see Michael's comment https://github.com/chapel-lang/chapel/issues/19050#issuecomment-1375836834

Michael also opened #21343 which discusses some syntax adjustments for calls to make interface disambiguation possible.

**Comment by mppf** [comment]
I've created a related issue, #21431, that focuses on how this approach can be achieved without stabilizing all of the constrained generics support.

**Comment by bradcray** [comment]
I'm closing this for reasons similar to why I closed #21431.
